lib/matplotlib/backends/backend_gtk3.py

```diff
@@ -211,10 +220,17 @@ def _get_key(self, event):
                     key = f'{prefix}+{key}'
         return key
 
+    def _update_device_pixel_ratio(self, instance, param):
+        # We need to be careful in cases with mixed resolution displays if
+        # device_pixel_ratio changes.
+        if self._set_device_pixel_ratio(self.get_scale_factor()):
+            self.draw()
```

Comment:
At least when testing with GDK_SCALE=2, this apparently needs to be called explicitly at some point, else the dpi is never set to 2.  Although the scaling also seems messed up even then...  (Haven't fully investigated yet, just a quick try.)

Reply:
My GTK/GDK doesn't seem to care about `GDK_SCALE`. I originally had this called in map-event as well; maybe that's needed for you?

Reply:
Ah, that environment variable only affects the x11 backend, not Wayland. To fix, it seems enough to trigger the pixel ratio update in `map-event`, which seems to require adding `STRUCTURE_MASK` to the event mask. Strangely, we get `configure-event` signals even though the documentation says that also needs the structure mask, so we could put the pixel ratio update there also.

This all seems to be unnecessary on Wayland, which has multi-monitor HiDPI support, and presumably had to be sure that the scale factor was correct all the time.

Reply:
Actually, I found `screen-changed`, which seems the better signal for this.